scalafix-core/shared/src/main/scala/scalafix/internal/util/TypeSyntax.scala

```diff
@@ -1,138 +1,451 @@
+      Some(new TypeSyntax(stpe, ctx, shortenNames, pos, table).prettify())
+    } catch {
+      case NonFatal(e) =>
+        val msg = new StringWriter()
+        e.printStackTrace(new PrintWriter(msg))
+        ctx.config.reporter.error(msg.toString, pos)
+        None
+    }
+  }
 
-    val functionN: SymbolMatcher = SymbolMatcher.exact(
-      0.to(22).map(i => Symbol(s"_root_.scala.Function$i#")): _*
-    )
-    val tupleN: SymbolMatcher = SymbolMatcher.exact(
-      1.to(22).map(i => Symbol(s"_root_.scala.Tuple$i#")): _*
-    )
+  implicit class XtensionSymbolInformationProperties(
+      info: s.SymbolInformation) {
+    private def test(property: s.SymbolInformation.Property): Boolean =
+      (info.properties & property.value) != 0
+    def isVal: Boolean = test(p.VAL)
+    def isVar: Boolean = test(p.VAR)
+    def isVarSetter: Boolean =
+      isVar && info.name.endsWith("_=")
+  }
 
-    def isStable(symbol: Symbol): Boolean = {
-      def loop(symbol: Symbol): Boolean = symbol match {
-        case Symbol.None => true
-        case Symbol.Global(owner, Signature.Term(_)) => loop(owner)
-        case _ => false
-      }
-      symbol match {
-        case Symbol.Global(owner, Signature.Term(_) | Signature.Type(_)) =>
-          loop(owner)
+  object T {
+    abstract class TypeRefExtractor(sym: String) {
+      def matches(tpe: s.Type): Boolean = tpe.tag match {
+        case t.TYPE_REF =>
+          tpe.typeRef.exists(_.symbol == sym)
+        case t.WITH_TYPE =>
+          tpe.withType.exists { x =>
+            x.types.lengthCompare(1) == 0 &&
+            unapply(x.types.head)
+          }
         case _ =>
           false
       }
+      def unapply(tpe: s.Type): Boolean = matches(tpe)
+    }
+    object AnyRef extends TypeRefExtractor("scala.AnyRef#")
+    object Nothing extends TypeRefExtractor("scala.Nothing#")
+    object Any extends TypeRefExtractor("scala.Any#")
+  }
+
+  def isFunctionN(symbol: String): Boolean = {
+    symbol.startsWith("scala.Function") &&
+    symbol.endsWith("#")
+  }
+
+  def isTupleN(symbol: String): Boolean = {
+    symbol.startsWith("scala.Tuple") &&
+    symbol.endsWith("#")
+  }
+
+  object FunctionN {
+    def unapply(symbol: String): Boolean = isFunctionN(symbol)
+  }
+  object TupleN {
+    def unapply(symbol: String): Boolean = isTupleN(symbol)
+  }
+}
+
+class TypeSyntax private (
+    rtpe: s.Type,
+    ctx: RuleCtx,
+    shortenNames: Boolean,
+    pos: Position,
+    _table: SymbolTable
+)(implicit index: SemanticdbIndex) {
+  val table = new CombinedSymbolTable(index, _table)
+
+  def fail(any: GeneratedMessage): Nothing = sys.error(any.toProtoString)
+
+  def isStable(symbol: Symbol): Boolean = {
+    def loop(symbol: Symbol): Boolean = symbol match {
+      case Symbol.None => true
+      case Symbol.Global(owner, Signature.Term(_)) => loop(owner)
+      case _ => false
     }
+    symbol match {
+      case Symbol.Global(owner, Signature.Term(_) | Signature.Type(_)) =>
```

Comment:
I read this as "A symbol is stable only if it is A Type or Term owned by Symbol.None or owned by a chain of n Global Terms".  Is that correct?

Reply:
This method suffers from the same problems as symbolToTree, it can be ignored.

Reply:
With the introduction of SymbolInformation one should no longer infer any information from the structure of symbols. Symbols are purely an ID and nothing more.